src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -0,0 +1,315 @@
+    this.collection = collection;
+  }
+
+  async readStream<EventType extends Event>(
+    streamName: StreamName,
+    options?: ReadStreamOptions<number>,
+  ): Promise<Exclude<ReadStreamResult<EventType, number>, null>> {
+    const expectedStreamVersion = options?.expectedStreamVersion;
+    const stream = await this.collection.findOne({
+      streamName: { $eq: streamName },
+    });
+
+    if (!stream) {
+      return {
+        events: [],
+        currentStreamVersion: MongoDBEventStoreDefaultStreamVersion,
+        streamExists: false,
+      };
+    }
+
+    assertExpectedVersionMatchesCurrent(
+      stream.events.length,
+      expectedStreamVersion,
+      MongoDBEventStoreDefaultStreamVersion,
+    );
+
+    const formattedEvents = stream.events
+      .slice(0, maxEventIndex(expectedStreamVersion))
+      .map(this.parseEvent<EventType>(streamName));
+
+    return {
+      events: formattedEvents,
+      currentStreamVersion: stream.events.length,
+      streamExists: true,
+    };
+  }
+
+  async aggregateStream<State, EventType extends Event>(
+    streamName: StreamName,
+    options: AggregateStreamOptions<State, EventType, number>,
+  ): Promise<AggregateStreamResult<State, number>> {
+    const stream = await this.readStream<EventType>(streamName, options?.read);
+    const state = stream.events.reduce(options.evolve, options.initialState());
+    return {
+      state,
+      currentStreamVersion: stream.currentStreamVersion,
+      streamExists: stream.streamExists,
+    };
+  }
+
+  async appendToStream<EventType extends Event>(
+    streamName: StreamName,
+    events: EventType[],
+    options?: AppendToStreamOptions<number> & {
+      /**
+       * These will be ran after a the events have been successfully appended to
+       * the stream. `appendToStream` will return after every projection is completed.
+       */
+      projections?: Array<
+        (stream: StreamToProject<EventType>) => void | Promise<void>
+      >;
+      /**
+       * Same as `options.projections` but this will run asynchronously.
+       */
+      asyncProjections?: Array<
+        (stream: StreamToProject<EventType>) => void | Promise<void>
+      >;
+    },
+  ): Promise<AppendToStreamResult<number>> {
+    const eventCreateInputs = events.map(this.stringifyEvent);
+
+    let stream = await this.collection.findOne({
+      streamName: { $eq: streamName },
+    });
+    let createdNewStream = false;
+
+    if (!stream) {
+      const result = await this.collection.insertOne({
+        streamName,
+        events: [],
+        createdAt: new Date(),
+        updatedAt: new Date(),
+      });
+      stream = await this.collection.findOne({
+        _id: result.insertedId,
+      });
+      createdNewStream = true;
+    }
+
+    // TODO: better error here, should rarely happen if ever
+    // if another error was not thrown before this
+    if (!stream) throw new Error('Failed to create stream');
```

Comment:
**QUESTION:** @alex-laycalvert what's the expected behaviour in terms of eventual consistency, read-your-own-writes guarantees and retries? Technically, with Mongo, if we don't provide retries and we're not using write or read concerns, then reading right after the write may not show the same document, which can cause issues. I described those issues longer in https://event-driven.io/en/dealing_with_eventual_consistency_and_idempotency_in_mongodb_projections/

Did you already think about it? I think that we'll need to have a logic described as in the article in the end. (this can come in the follow up implementation tho).

Reply:
I have not thought about it, I'll definitely take a look at the article.

Reply:
@alex-laycalvert ok, cool, I'm curious your thoughts on that. Still, that's something that we can/should do in a dedicated PR.

Reply:
Ah,and @alex-laycalvert the code I described in article, you can find here: 
- https://github.com/oskardudycz/EventSourcing.NodeJS/blob/0a08b4237b573747df7bacbf9ee195c4e8970edf/samples/webinar/src/core/mongoDB.ts#L102
- https://github.com/oskardudycz/EventSourcing.NodeJS/blob/0a08b4237b573747df7bacbf9ee195c4e8970edf/samples/webinar/src/shoppingCarts/shoppingCartDetails.ts#L104

Feel free to use it :)